core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala

```diff
@@ -179,8 +198,13 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
             blocks, SparkEnv.get.serializer, compressionCodec)
           // Store the merged copy in BlockManager so other tasks on this executor don't
           // need to re-fetch it.
-          SparkEnv.get.blockManager.putSingle(
-            broadcastId, obj, StorageLevel.MEMORY_AND_DISK, tellMaster = false)
+          blockManager.putSingle(broadcastId, obj, StorageLevel.MEMORY_AND_DISK, tellMaster = false)
+          Option(TaskContext.get()) match {
+            case Some(taskContext) =>
+              taskContext.addTaskCompletionListener(_ => blockManager.releaseLock(broadcastId))
+            case None =>
+              blockManager.releaseLock(broadcastId)
```

Comment:
This branch corresponds to a case where you call `putSingle` outside of the context of a task. If we don't release this lock here then the driver will never be able to delete its local copy of broadcast variables. We can't have `put()` automatically release locks because that's going to introduce race conditions in the `CacheManager` simplification patch.